transactionoutbox-core/src/main/java/com/gruelbox/transactionoutbox/DefaultPersistor.java

```diff
@@ -248,6 +256,13 @@ public boolean unblock(Transaction tx, String entryId) throws Exception {
   @Override
   public List<TransactionOutboxEntry> selectBatch(Transaction tx, int batchSize, Instant now)
       throws Exception {
+    List<TransactionOutboxEntry> results = selectBatchUnordered(tx, batchSize, now);
```

Comment:
We have a problem here I think. When `flush()` runs, it doesn't run the items it loads; it _submits_ them. That means that even if it submits the work to run in the right order, those threads might not _run_ in the submitted order.  FIrstly because they might "race", but more likely because one might fail, causing it to go back on the queue and get run later.

We need to run one entry at a time, synchronously, and if that fails, _keep retrying it forever_.

That means:

 * We have to handle ordered processing completely separately from unordered processing.
 * No calling `Submitter` in `flush()` when dealing with ordered entrues. Simply call `processNow()`

As for the most efficient way to fetch the data given that we need to only process one-at-a-time for each `groupId`, there are few different approaches:

 * EITHER Don't fetch ordered records in batches. Fetch them one at a time for each group. You then have to block until they _all_ complete before you can process the next batch, but you _can_process all of these in parallel safely (using a fork/join behaviour - easy with `CompletableFuture`)
```
SELECT * FROM TXNO_OUTBOX a
WHERE groupId <> NULL 
AND sequence =
 (SELECT MIN(sequence) FROM TXNO_OUTBOX b
  WHERE a.groupId = b.groupId
  AND processed = false
  AND blocked = false)
``` 
* OR you could first fetch a list of all the distinct group ids and spin up independent parallel processors for each which simply fetch one record at a time.
* More ideas welcome of course!

Reply:
Thanks for the comments. So separating flush for ordered processing and invoking `processNow` seems to be the way to go here. For failed tasks, it seems that as long as tasks are processed in order within the group and we keep retrying the first task in the group until it succeeds we should be fine.

Reply:
It would be nice to merge https://github.com/gruelbox/transaction-outbox/pull/484 (maybe not necessarily as-is currently, but the idea seems solid) as I believe that would make the code easier to work with, especially when introducing additional SQL as in this PR.

It might also be nice to merge https://github.com/gruelbox/transaction-outbox/pull/492 in case sequences are needed directly as the "create sequence" functionality is improved in later versions.

Reply:
Updated the branch to process ordered tasks synchronously (while unordered tasks are still submitted). Also added a test to ensure that a blocked ordered task will prevent other tasks in the same ordered group from progressing.